model/task_blocked.go

```diff
@@ -0,0 +1,252 @@
+package model
+
+import (
+	"strings"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/build"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"github.com/tychoish/tarjan"
+)
+
+const (
+	taskBlocked  = "blocked"
+	taskPending  = "pending"
+	taskRunnable = "runnable"
+)
+
+// BlockedState returns "blocked," "pending" (unsatisfied dependencies,
+// but unblocked), or "" (runnable) to represent the state of the task
+// with respect to its dependencies
+func BlockedState(t *task.Task, tasksWithDeps []task.Task) (string, error) {
+	if t.DisplayOnly {
+		return blockedStateForDisplayTask(t, tasksWithDeps)
+	}
+	if t.IsPartOfSingleHostTaskGroup() {
+		return blockedStateForTaskGroups(t)
+	}
+	return blockedStatePrivate(t)
+}
+
+func blockedStatePrivate(t *task.Task) (string, error) {
+	if len(t.DependsOn) == 0 {
+		return taskRunnable, nil
+	}
+	dependencyIDs := []string{}
+	for _, d := range t.DependsOn {
+		dependencyIDs = append(dependencyIDs, d.TaskId)
+	}
+	dependentTasks, err := task.Find(task.ByIds(dependencyIDs).WithFields(task.DisplayNameKey, task.StatusKey,
+		task.ActivatedKey, task.BuildVariantKey, task.DetailsKey, task.DependsOnKey))
+	if err != nil {
+		return "", errors.Wrap(err, "error finding dependencies")
+	}
+	taskMap := map[string]*task.Task{}
+	for i := range dependentTasks {
+		taskMap[dependentTasks[i].Id] = &dependentTasks[i]
+	}
+	for _, dependency := range t.DependsOn {
+		depTask := taskMap[dependency.TaskId]
+		state, err := getStateByDependency(depTask, dependency)
+		if err != nil {
+			return "", errors.Wrap(err, "")
+		}
+		if state != taskRunnable {
+			return state, nil
+		}
+	}
+	return taskRunnable, nil
+}
+
+// getStateByDependency determines if the task is still running, is blocked, or has violated the given dependency
+func getStateByDependency(t *task.Task, dependency task.Dependency) (string, error) {
+	if t == nil {
+		grip.Error(message.Fields{
+			"message": "task does not exist",
+			"task_id": dependency.TaskId,
+		})
+		return taskRunnable, nil
+	}
+	state, err := blockedStatePrivate(t)
+	if err != nil {
+		return "", err
+	}
+	if state == taskBlocked {
+		return taskBlocked, nil
+	} else if t.Status == evergreen.TaskSucceeded || t.Status == evergreen.TaskFailed {
+		if t.Status != dependency.Status && dependency.Status != AllStatuses {
+			return taskBlocked, nil
+		}
+	} else {
+		return taskPending, nil
+	}
+	return taskRunnable, nil
+}
+
+func blockedStateForTaskGroups(t *task.Task) (string, error) {
+	tasks, err := GetTasksInTaskGroup(t)
+	if err != nil {
+		return "", errors.Wrap(err, "problem getting tasks in task group")
+	}
+
+	// get dependencies from all tasks in group
+	dependsOn := []task.Dependency{}
+	dependencyIDs := []string{}
+	for _, curTask := range tasks {
```

Comment:
I think this only finds dependencies, and their dependencies. It's not actually recursive. Could you use the same logic as `blockedStateForDisplayTask`, with the only difference between finding task group tasks instead of exec tasks on the first line? I realize that requires a different approach in https://github.com/evergreen-ci/evergreen/pull/2368/files#diff-bc9280b69845be379dfa476eb5c333d5R28, but I think it's worth it for simplicity and to eliminate two different approaches to what seem to be essentially the same problem.